### PR TITLE
UE ip pool allocation by UPF

### PR DESCRIPTION
**Pull request by badhrinathpa**
Changes include : 
1. config for ip pool and ip pool feature flag
2. send UPF features flag in Association response with UEIP flag.
3. alloc UE ip in case when upf supports this feature and smf has sent IP with only CHV4 flag.
4. send PDR with UE info in establishment response

### CODE REVIEW

File: pfcpiface/ip_pool.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright(c) 2020 Intel Corporation
```
Comment (thakurajayL) [inline pfcpiface/ip_pool.go:2]: Please add ONF copyright header.

File: pfcpiface/session-pdr.go [hunk 2]
```diff
@@ -5,8 +5,43 @@ package main
...
 )
 
+// Release allocated IPs
+func releaseAllocatedIPs(upf *upf, session *PFCPSession) {
+	log.Println("release allocated IPs")
+	for _, pdr := range session.pdrs {
+		if (pdr.allocIpFlag) && (pdr.srcIface == core) {
+			log.Println("pdrID : ", pdr.pdrID)
+			var ueIP net.IP = int2ip(pdr.dstIP)
+			log.Println("ueIP : ", ueIP.String())
```
Comment (thakurajayL) [inline pfcpiface/session-pdr.go:21]: Clubbing 2 logs would help.

File: pfcpiface/utils.go [hunk 3]
```diff
@@ -23,6 +23,28 @@ func Set(b, flag Bits) Bits { return b | flag }
...
 //func Toggle(b, flag Bits) Bits { return b ^ flag }
 //func Has(b, flag Bits) bool { return b&flag != 0 }
+func setUeipFeature(features ...uint8) {
+	if len(features) >= 3 {
+		features[2] = features[2] | 0x04
+	}
+}
+
+func has2ndBit(f uint8) bool {
+	return (f&0x02)>>1 == 1
```
Comment (thakurajayL) [inline pfcpiface/utils.go:33]: if bool return is not required then we can avoid >> 1 & == 1 check.
return (f & 0x02)

**Comment by thakurajayL** [comment]
In general changes look good.

**Comment by krsna1729** [comment]
@badhrinathpa @thakurajayL what do you think of using something like this for IPAM? 
https://github.com/metal-stack/go-ipam

File: pfcpiface/bess.go [hunk 4]
```diff
@@ -278,6 +278,12 @@ func (b *bess) setUpfInfo(u *upf, conf *Conf) {
 		u.simInfo = simInfo
 	}
 
+	u.ippool_cidr = conf.CPIface.UeIPPool
```
Comment (krsna1729) [inline pfcpiface/bess.go:281]: change all naming to not have underscore. I thought one of the linting tool checks and throws warning

File: pfcpiface/main.go [hunk 5]
```diff
@@ -52,9 +52,11 @@ type SimModeInfo struct {
 
 // CPIfaceInfo : CPIface interface settings
 type CPIfaceInfo struct {
-	DestIP   string `json:"nb_dst_ip"`
-	SrcIP    string `json:"nb_src_ip"`
-	FQDNHost string `json:"hostname"`
+	EnableUeIPAlloc bool   `json:"enable_ue_ip_alloc"`
```
Comment (krsna1729) [inline pfcpiface/main.go:55]: keep this before UE IP pool

File: pfcpiface/ip_pool.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+type ip_pool struct {
+	free_pool []string
+}
+
+func needAllocIp(ueIPaddr *ie.UEIPAddressFields) bool {
+	if has2ndBit(ueIPaddr.Flags) && !has5thBit(ueIPaddr.Flags) {
+		return false
+	}
+	return true
+}
```
Comment (krsna1729) [inline pfcpiface/ip_pool.go:23]: move this to parse-pdr.go itself? that way this file will purely be IP allocation related.

File: pfcpiface/parse-pdr.go [hunk 7]
```diff
@@ -212,7 +227,7 @@ func (p *pdr) parsePDI(pdiIEs []*ie.IE, appPFDs map[string]appPFD) error {
 	return nil
 }
 
-func (p *pdr) parsePDR(ie1 *ie.IE, seid uint64, appPFDs map[string]appPFD) error {
+func (p *pdr) parsePDR(ie1 *ie.IE, seid uint64, appPFDs map[string]appPFD, upf *upf) error {
```
Comment (krsna1729) [inline pfcpiface/parse-pdr.go:230]: Do we need to pass entire UPF object or just ip_pool struct is enough?

File: pfcpiface/ip_pool.go (hunk 1 above)
Comment (badhrinathpa) [reply pfcpiface/ip_pool.go:2]: done

File: pfcpiface/session-pdr.go (hunk 2 above)
Comment (badhrinathpa) [reply pfcpiface/session-pdr.go:21]: done

File: pfcpiface/utils.go (hunk 3 above)
Comment (badhrinathpa) [reply pfcpiface/utils.go:33]: we would have to add check for ==1 wherever we call has2ndBit. So will let it be as it is for now.

File: pfcpiface/bess.go (hunk 4 above)
Comment (badhrinathpa) [reply pfcpiface/bess.go:281]: done

File: pfcpiface/main.go (hunk 5 above)
Comment (badhrinathpa) [reply pfcpiface/main.go]: done

File: pfcpiface/ip_pool.go (hunk 6 above)
Comment (badhrinathpa) [reply pfcpiface/ip_pool.go]: done

File: pfcpiface/parse-pdr.go (hunk 7 above)
Comment (badhrinathpa) [reply pfcpiface/parse-pdr.go:237]: its just pointer passing. might need other values later. So keeping as it is.

Comment (krsna1729) [reply pfcpiface/parse-pdr.go:237]: I was thinking in terms layer/scope instead pointer or not.

**Review by krsna1729 — APPROVED** [review]

**Comment by yoooou** [comment]
retest this please

**Pull request merged**